Automattic/jetpack
# Jetpack core api: Refactor and reuse trait for proxying requests to WPCOM

This PR refactors the existing `WPCOM_REST_API_Proxy_Request_Trait` so that it can be used by any shared endpoint (between self-hosted and Simple sites) that needs to forward the request to WPCOM when it runs on the Jetpack end.

It also refactors two existing endpoints to use the trait and introduces unit tests for both:
- `WP_Test_WPCOM_REST_API_V3_Endpoint_Blogging_Prompts`
- `WPCOM_REST_API_V2_Endpoint_Memberships`

## Proposed changes:

**`WPCOM_REST_API_Proxy_Request_Trait`:** 
- Introduce `proxy_request_to_wpcom` to allow falling back to using the Site-level Connection if the current user is not connected
- Update `proxy_request_to_wpcom_as_user` for proxying requests to WPCOM on behalf of the current user
- Introduce `proxy_request_to_wpcom_as_blog` for proxying requests to WPCOM using the Site-level Connection

**`WP_Test_WPCOM_REST_API_V3_Endpoint_Blogging_Prompts`:** 
- Refactor endpoint to use the `WPCOM_REST_API_Proxy_Request_Trait` trait
- Add unit tests

**`WPCOM_REST_API_V2_Endpoint_Memberships`:** 
- Refactor endpoint to use the `WPCOM_REST_API_Proxy_Request_Trait` trait
- Add unit tests

### Other information:


## Jetpack product discussion


p1730895681631659-slack-C05PV073SG3

## Does this pull request change what data or activity we track or use?


No

## Testing instructions:






* Make sure to run the following command on your sandbox: `bin/jetpack-downloader test jetpack update/forward-wpcom-request-trait`
* We'll need to test the [Payments block](https://jetpack.com/support/jetpack-blocks/payments-block/) and the [Writing Prompt Block](https://jetpack.com/support/jetpack-blocks/writing-prompt-block/) across all environments: **self-hosted, WoA and Simple**
* Add a Donations Form block in a post and play around with all the related fields and block settings. 
* Confirm behaviour same as trunk

<img width="1889" alt="Screenshot 2024-11-20 at 13 09 01" src="https://github.com/user-attachments/assets/8355a0eb-e6e6-4439-814b-f43e275b8f3c">

* Add a Writing Prompt block in a post and confirm behaviour same as trunk

<img width="970" alt="Screenshot 2024-11-20 at 13 17 23" src="https://github.com/user-attachments/assets/270abb15-b9d2-440e-8840-6b026dea3625">
